MC/Core/Src/freertos.c

```diff
@@ -0,0 +1,515 @@
+
+		osMessageQueueGet(start_frame_q, &start_frame, NULL, osWaitForever);
+
+		//Process data from MC
+		mc_data.direction = bits_to_direction(CAN_get_segment(start_frame, DRIVING_DIRECTION_K));
+		mc_data.speed = CAN_get_segment(start_frame, MOTOR_SPEED_K);
+		mc_data.error_code = error_code(CAN_get_segment(start_frame, MOTOR_ERROR_CODE_K));
+
+		//Create new CAN_frame with processed data
+		CAN_set_segment(&send_frame, DRIVING_DIRECTION_K, mc_data.direction);
+		CAN_set_segment(&send_frame, MOTOR_SPEED_K, mc_data.speed);
+		CAN_set_segment(&send_frame, MOTOR_ERROR_CODE_K, mc_data.error_code);
+
+	    //Send data to RPI
+	    if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan1)) {
+	        CAN_send_frame(send_frame);
+	    }
+
+
+	    osDelay(1);
+
+
+	}
+}
+
+void MC_SendEndThread(void *argument){
+
+	CAN_Frame_t end_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER_K2); //Initialize CAN_frame to send to MC
+	CAN_Frame_t send_frame = CAN_frame_init(&hcan1, MOTOR_CONTROLLER_K2); //Initialize CAN_frame to send to RPI
+
+	Motor_Controller_t mc_data = {0};
+
+	for(;;){
+
+		osSemaphoreAcquire(MCSemEndHandle, osWaitForever);
+
+		osMessageQueueGet(end_frame_q, &end_frame, NULL, osWaitForever);
+
+		//Process data from MC
+		mc_data.voltage = process_bits(CAN_get_segment(end_frame, BATTERY_VOLTAGE_K));
+		mc_data.current = process_bits(CAN_get_segment(end_frame, BATTERY_CURRENT_K));
+		mc_data.motor_temp = process_bits(CAN_get_segment(end_frame, MOTOR_TEMP_K));
+		mc_data.controller_temp = process_bits(CAN_get_segment(end_frame, MOTOR_CONTROLLER_TEMP_K));
+
+		//Create new CAN_frame with processed data
+		CAN_set_segment(&send_frame, BATTERY_VOLTAGE_K, mc_data.voltage);
+		CAN_set_segment(&send_frame, BATTERY_CURRENT_K, mc_data.current);
+		CAN_set_segment(&send_frame, MOTOR_TEMP_K, mc_data.motor_temp);
+		CAN_set_segment(&send_frame, MOTOR_CONTROLLER_TEMP_K, mc_data.controller_temp);
+
+		//Send data to RPI
+	    if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan1)) {
+	    	CAN_send_frame(send_frame);
+	    }
+
+	    osDelay(1);
+
+
+	}
+}
+
+void MC_CommandControlThread(void *argument){
+	CAN_Frame_t command_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER); //initialize frame to receive command from RPI (psuedo frame, RPI code doesn't exist yet)
+	Motor_Controller_t kelly_mc =  MC_init(&hcan3, &hi2c2);; //Initialize motor controller with used CAN and I2C handles
+
+	for(;;){
+
+		//Checks if CAN message is available from RPI
+		if(HAL_CAN_GetRxFifoFillLevel(&hcan3, CAN_RX_FIFO1)){
+			command_frame = CAN_get_frame(&hcan3, CAN_RX_FIFO1);
+		}
+
+		uint8_t command = CAN_get_segment(command_frame, RPI_COMMAND_CODE); //Get "command" from CAN data frame
+
+		if(command == MC_START){
+			MC_drive(&kelly_mc);
+		}
+		else if(command == MC_STOP){
+			MC_stop(&kelly_mc);
+		}
+		else if(command == MC_THROTTLE){
+			uint8_t throttle = CAN_get_segment(command_frame, RPI_COMMAND_DATA); //Get throttle percent "data" from CAN data frame
```

Comment:
needs more processing I assume once we have real CAN data from RPI set up

Reply:
yes we'll see how to process the data once we get the actual rpi setup